### PR TITLE
html kernel can append to container

### PR DESCRIPTION
**Pull request by colombod**
html kernel can be configured to reuse a specific container and add to it content or replace it

### CODE REVIEW

File: src/microsoft-dotnet-interactive/src/htmlKernel.ts [hunk 1]
```diff
@@ -87,20 +87,43 @@ export function htmlDomFragmentInserter(htmlFragment: string, configuration?: Ht
...
-        (htmlFragment: string) => htmlDomFragmentInserter(htmlFragment, {
-            getOrCreateContainer: () => config.container,
-            updateContainerContent: (container, htmlFragment) => {
-                if (config.contentBehaviour === "append") {
-                    container.innerHTML += htmlFragment;
-                } else {
-                    container.innerHTML = htmlFragment;
+export type HtmlKernelInBrowserConfiguration = { kernelName: string, container: HTMLElement | string, contentBehaviour: "append" | "replace" } | { kernelName: string, htmlDomFragmentInserterConfiguration: HtmlDomFragmentInserterConfiguration };
+
+export function createHtmlKernelThatWorksWithPageDomInBrowser(config: HtmlKernelInBrowserConfiguration): Kernel {
```
Comment (jonsequitur) [inline src/microsoft-dotnet-interactive/src/htmlKernel.ts:92]: This name is a little verbose if we want to make it public.

Comment (colombod) [reply src/microsoft-dotnet-interactive/src/htmlKernel.ts:92]: something like `createHtmlKernelForBrowserHoating`?

Comment (jonsequitur) [reply src/microsoft-dotnet-interactive/src/htmlKernel.ts]: I think even `createHtmlKernelForBrowser` gets the point across well.

File: src/microsoft-dotnet-interactive/src/htmlKernel.ts [hunk 2]
```diff
@@ -20,62 +20,110 @@ export class HtmlKernel extends Kernel {
...
-    containerFactory?: () => HTMLDivElement,
-    elementToObserve?: () => HTMLElement,
-    addToDom?: (element: HTMLElement) => void,
-    mutationObserverFactory?: (callback: MutationCallback) => MutationObserver
-}): Promise<void> {
+export interface HtmlDomFragmentInserterConfiguration {
+    getOrCreateContainer?: () => HTMLElement,
+    updateContainerContent?: (container: HTMLElement, htmlFragment: string) => void,
+    createMutationObserver?: (callback: MutationCallback) => MutationObserver,
+    nomarliseHtmlFragment?: (htmlFragment: string) => string
```
Comment (jonsequitur) [inline src/microsoft-dotnet-interactive/src/htmlKernel.ts:48]:
```suggestion
    normalizeHtmlFragment?: (htmlFragment: string) => string
```

File: src/microsoft-dotnet-interactive/src/htmlKernel.ts [hunk 3]
```diff
@@ -6,10 +6,10 @@ import { Kernel, IKernelCommandInvocation } from "./kernel";
 import { PromiseCompletionSource } from "./promiseCompletionSource";
 
 export class HtmlKernel extends Kernel {
-    constructor(kernelName?: string, private readonly htmlFragmentProcessor?: (htmlFragment: string) => Promise<void>, languageName?: string, languageVersion?: string) {
+    constructor(kernelName?: string, private readonly htmlFragmentInserter?: (htmlFragment: string) => Promise<string>, languageName?: string, languageVersion?: string) {
         super(kernelName ?? "html", languageName ?? "HTML");
-        if (!this.htmlFragmentProcessor) {
-            this.htmlFragmentProcessor = domHtmlFragmentProcessor;
+        if (!this.htmlFragmentInserter) {
+            this.htmlFragmentInserter = htmlDomFragmentInserter;
```
Comment (jonsequitur) [inline src/microsoft-dotnet-interactive/src/htmlKernel.ts:12]:
```suggestion
            this.htmlFragmentInserter = htmlFragmentInserter;
```

File: src/microsoft-dotnet-interactive/src/htmlKernel.ts (hunk 1 above)
Comment (colombod) [reply src/microsoft-dotnet-interactive/src/htmlKernel.ts]: thanks!

**Review by jonsequitur — APPROVED** [review]

**Pull request merged**